vm/actor/src/builtin/miner/mod.rs

```diff
@@ -667,6 +669,191 @@ impl Actor {
+                deal_ids: precommit.info.deal_ids.clone(),
+                sector_type: precommit.info.seal_proof,
+            });
+        }
+
+        let comm_ds = request_unsealed_sector_cids(rt, &compute_data_commitments_inputs)?;
+        let mut svis = Vec::new();
+        let miner_actor_id: u64 = if let Payload::ID(i) = rt.message().receiver().payload() {
+            *i
+        } else {
+            return Err(actor_error!(
+                ErrIllegalState,
+                "runtime provided non-ID receiver address {}",
+                rt.message().receiver()
+            ));
+        };
+        let receiver_bytes = rt.message().receiver().marshal_cbor().map_err(|e| {
+            ActorError::from(e).wrap("failed to marshal address for seal verification challenge")
+        })?;
+
+        for (i, precommit) in precommits.iter().enumerate() {
+            let interactive_epoch = precommit.pre_commit_epoch + PRE_COMMIT_CHALLENGE_DELAY;
+            if rt.curr_epoch() <= interactive_epoch {
+                return Err(actor_error!(
+                    ErrForbidden,
+                    "too early to prove sector {}",
+                    precommit.info.sector_number
+                ));
+            }
+            let sv_info_randomness = rt.get_randomness_from_tickets(
+                SealRandomness,
+                precommit.info.seal_rand_epoch,
+                &receiver_bytes,
+            )?;
+            let sv_info_interactive_randomness = rt.get_randomness_from_beacon(
+                InteractiveSealChallengeSeed,
+                interactive_epoch,
+                &receiver_bytes,
+            )?;
+            let svi = AggregateSealVerifyInfo {
+                sector_number: precommit.info.sector_number,
+                randomness: sv_info_randomness,
+                interactive_randomness: sv_info_interactive_randomness,
+                sealed_cid: precommit.info.sealed_cid,
+                unsealed_cid: comm_ds[i],
+            };
+            svis.push(svi);
+        }
+
+        let seal_proof = precommits[0].info.seal_proof;
+        if precommits.is_empty() {
+            return Err(actor_error!(
+                ErrIllegalState,
+                "bitfield non-empty but zero precommits read from state"
+            ));
+        }
+        rt.verify_aggregate_seals(&AggregateSealVerifyProofAndInfos {
+            miner: miner_actor_id,
+            seal_proof,
+            aggregate_proof: fil_types::RegisteredAggregateProof::SnarkPackV1,
+            proof: params.aggregate_proof,
+            infos: svis,
+        })
+        .map_err(|e| {
+            e.downcast_default(ExitCode::ErrIllegalArgument, "aggregate seal verify failed")
+        })?;
+        confirm_sector_proofs_valid_internal(rt, precommits_to_confirm.clone())?;
+        // Compute and burn the aggregate network fee. We need to re-load the state as
+        // confirmSectorProofsValid can change it.
+        let state: State = rt.state()?;
```

Comment:
Could you update #1105 with where you got this code so I can review against that also? I have a feeling there's code referenced here that's not in Lotus 1381.